libcxx/test/std/utilities/tuple/tuple.tuple/tuple.apply/make_from_tuple.verify.cpp

```diff
@@ -0,0 +1,46 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// UNSUPPORTED: c++03, c++11, c++14
```

Comment:
Can you make the test a SFINAE based test too, then we don't need to verify the compiler output.

Reply:
Agree, but current implementation in this PR seems not SFINAE friendly, there some hard errors. I'm trying to solve these problems. 
```
template <class T, class Tuple, class = std::void_t<>>
struct can_make_from_tuple : std::false_type {};

template <class T, class Tuple>
struct can_make_from_tuple<
    T,
    Tuple,
    std::void_t<decltype(std::make_from_tuple<T>(std::declval<Tuple>()))>>
    : std::true_type {};

static_assert(!can_make_from_tuple<char*, std::tuple<const char*>>::value);
```

Reply:
I was wondering if we can also add a same constraints `is_constructible_v<_Tp, decltype(std::get<_Idx>(std::forward<_Tuple>(__t)))...>` to `std::make_from_tuple`.

Reply:
Possible implementation:
```
#if _LIBCPP_STD_VER >= 20
template <class _Tp, class _Tuple, size_t... _Idx>
inline _LIBCPP_HIDE_FROM_ABI constexpr _Tp __make_from_tuple_impl(_Tuple&& __t, __tuple_indices<_Idx...>)
  noexcept(noexcept(_Tp(std::get<_Idx>(std::forward<_Tuple>(__t))...)))
  requires is_constructible_v<_Tp, decltype(std::get<_Idx>(std::forward<_Tuple>(__t)))...> {
  return _Tp(std::get<_Idx>(std::forward<_Tuple>(__t))...);
}
#else
template <class _Tp, class _Tuple, size_t... _Idx>
inline _LIBCPP_HIDE_FROM_ABI constexpr _Tp __make_from_tuple_impl(_Tuple&& __t, __tuple_indices<_Idx...>, 
    enable_if_t<is_constructible_v<_Tp, decltype(std::get<_Idx>(std::forward<_Tuple>(__t)))...>> * = nullptr)
    _LIBCPP_NOEXCEPT_RETURN(_Tp(std::get<_Idx>(std::forward<_Tuple>(__t))...))
#endif // _LIBCPP_STD_VER >= 20

// clang-format on
template <class _Tp, class _Tuple, class = std::void_t<>>
struct __can_make_from_tuple : std::false_type {};
template <class _Tp, class _Tuple>
struct __can_make_from_tuple<
    _Tp, _Tuple,
    std::void_t<decltype(std::__make_from_tuple_impl<_Tp>(
        std::declval<_Tuple>(),
        std::declval<typename __make_tuple_indices<
            tuple_size_v<remove_reference_t<_Tuple>>>::type>()))>>
    : std::true_type {};

template <class _Tp, class _Tuple>
inline _LIBCPP_HIDE_FROM_ABI constexpr _Tp make_from_tuple(
    _Tuple &&__t,
    std::enable_if_t<__can_make_from_tuple<_Tp, _Tuple>::value> * = nullptr)
    _LIBCPP_NOEXCEPT_RETURN(std::__make_from_tuple_impl<_Tp>(
        std::forward<_Tuple>(__t),
        typename __make_tuple_indices<
            tuple_size_v<remove_reference_t<_Tuple>>>::type{}))
// clang-format off
```

Reply:
> I was wondering if we can also add a same constraints `is_constructible_v<_Tp, decltype(std::get<_Idx>(std::forward<_Tuple>(__t)))...>` to `std::make_from_tuple`.

It seems to me that we should _only_ add constraints to `std::make_from_tuple` if SFINAE-friendliness is required. Whether the internal `__make_from_tuple_impl` is constrained shouldn't be observable.

Reply:
Note that it is not conforming to add an additional default argument to `std::make_from_tuple`, see [[global.functions]/3](https://eel.is/c++draft/global.functions#3).

Reply:
I've made a SFINAE based test to test `std::__make_from_tuple_impl` directly.